### PR TITLE
TxPool efficiency improvements

### PR DESCRIPTION
**Pull request by tbcd**
A bunch of little efficiency improvements for TxPool, mostly around avoiding iterations or allocations.

### CODE REVIEW

File: txpool/pool.go [hunk 1]
```diff
@@ -619,41 +625,47 @@ func (p *TxPool) validateTxs(txs TxSlots) (reasons []DiscardReason, goodTxs TxSl
...
+			p.punishSpammer(txn.senderID)
+		}
+		reasons[i] = reason
+	}
+
+	goodTxs.Resize(uint(goodCount))
+
+	j := 0
+	for i, txn := range txs.txs {
+		if reasons[i] == NotSet {
```
Comment (AlexeyAkhunov) [inline txpool/pool.go:645]: Should not this be `reasons[i] == Success?` Because now nothing is setting it to `NotSet` above

**Comment by AlexeyAkhunov** [comment]
Generally looks really good, thank you just one thing looks potentially wrong in `validateTxs` function (left comment)

File: txpool/pool.go (hunk 1 above)
Comment (tbcd) [reply txpool/pool.go:645]: I agree this one is a little unintuitive. `NotSet` is the default value when the `reasons` slice is created, and `validateTxs()` is expected to leave all successes as `NotSet` for the return value, so it only changes a value for a discard.

Of the callers to `validateTxs()`, `OnNewBlock()` and `processRemoteTxs()` both ignore `reasons`, which leaves the other caller, `AddLocalTxs()`.

`AddLocalTxs()` calls `fillDiscardReasons()`, which is the "final" filler -- it looks at only the `NotSet` txs to see if there's any other cause to discard them (`discardReasonsLRU` by `IdHash`), and otherwise sets the final `Success` state.

In other words, a txn is not `Success` until all checks are done, and until then it is inconclusive with `NotSet`.

**Comment by tbcd** [comment]
> Generally looks really good, thank you just one thing looks potentially wrong in `validateTxs` function (left comment)

No problem, I'll a comment that makes it clearer about the default `NotSet` state.

File: txpool/pool.go (hunk 1 above)
Comment (AlexeyAkhunov) [reply txpool/pool.go:648]: Thanks a lot! Now I see it
